### DISCUSSION TITLE
Incomplete translations

### DISCUSSION THREAD

**Discussion by martio** [laravel/nova-issues#3239] [Q&A] [answered]
- Laravel Version: 8.25.0
- Nova Version: 3.20.0
- PHP Version: 8.0.1

### Description:

Incomplete translation when localization is set to another language in the app config file.

### Detailed steps to reproduce the issue on a fresh Nova installation:

FILE: config/app.php

When _**'locale' => 'en'**_ is **OK**:

<img width="1440" alt="Zrzut ekranu 2021-02-10 o 11 16 03" src="https://user-images.githubusercontent.com/133636/107496743-fc2a6980-6b91-11eb-8122-96662fb81ca1.png">

When _**'locale' => 'pl'**_ is **INCOMPLETE** ("Resouces" label in sidebar and "Create" button):

<img width="1440" alt="Zrzut ekranu 2021-02-10 o 11 16 26" src="https://user-images.githubusercontent.com/133636/107496787-0c424900-6b92-11eb-8c08-5b134463b8ca.png">

NovaServiceProvider:

```
    public function boot()
    {
        parent::boot();

        Nova::serving(function (ServingNova $event) {
            $this->app->setLocale('pl');

            Nova::translations([
                resource_path('lang/vendor/nova/en.json'),
                resource_path('lang/vendor/nova/pl.json'),
            ]);
        });
    }
```

**Comment by crynobone** [comment]
![image](https://user-images.githubusercontent.com/172966/107502799-20a73580-6bd4-11eb-95f8-1e862f6d3660.png)

This doesn't seem to be correct.

**Comment by crynobone** [comment]
Based on code review, you shouldn't use the above code since Laravel Nova would already load `en.json` or `pl.json` based on locale config.

**Comment by f-liva** [comment]
It doesn't work for me either, in my case the local language is Italian (it).

![Immagine 2021-02-22 162957](https://user-images.githubusercontent.com/346224/108729913-5086ff00-752b-11eb-8659-8a7ccf99e5bf.png)
![Immagine 2021-02-22 163028](https://user-images.githubusercontent.com/346224/108729923-5250c280-752b-11eb-93cd-14ea0704bca0.png)

Red circles are in English, green circles are in Italian.

So strange...

**Reply by martio** [reply]
It will work as you set the default language to English:

```
'locale' => 'en',
```

in config/app.php

This is bug...

**Reply by crynobone** [reply]
Loading both `en.json` and `pl.json` causes the bug. You should only load language string based on the current locale.

**Comment by f-liva** [comment]
Nothing changes by setting all language references to IT instead of EN and importing only the IT language.

It still loads half in English and half in Italian.

**Reply by martio** [reply]
I have the same...

**Comment by crynobone** [comment]
![image](https://user-images.githubusercontent.com/172966/108822186-b2159f00-75f9-11eb-8cc8-df8dbc7a545d.png)
![image](https://user-images.githubusercontent.com/172966/108822217-bcd03400-75f9-11eb-9bc2-ce30341c332e.png)
![image](https://user-images.githubusercontent.com/172966/108822229-c063bb00-75f9-11eb-8803-b8c5406dafa7.png)

You can refer reproducing code https://github.com/nova-issues/issue-3239/commit/9bb0a0f2306678aa289bbd3e4c6a791caefc560e

**Reply by martio** [reply]
@crynobone it doesn't work for me...

```
    public function boot()
    {
        parent::boot();

        Nova::serving(function (ServingNova $event) {
            $this->app->setLocale((('pl' == $event->request->user()->locale) ? 'pl' : 'en'));

            Nova::translations(
                resource_path("lang/vendor/nova/{$this->app->getLocale()}.json")
            );
        });
    }
```

```
    'locale' => 'pl',
```

**Reply by martio** [reply]
It's okay when I set in app.php: 'locale' => 'en'

**Reply by crynobone** [reply]
above code the one causing the issue. See the example project I share above

**Comment by f-liva** [comment]
@crynobone I didn't understand your request

**Reply by henryavila** [reply]
What he means, is that the code you added in `NovaServiceProvider` file manually adding the translation is breaking the app, because nova will automatically handle the translation. See my comment below for more detail.

Also, take a look in docs: https://nova.laravel.com/docs/3.0/customization/localization.html

**Comment by henryavila** [comment, marked as answer]
The translation is working perfect for me. I use the entire app in 'pt_BR' locale.
You have to do just 3 things:

1) Get the json lang file. I recommend using: https://github.com/coderello/laravel-nova-lang and insert your custom content
2) Put the `.json` lang file in `resources/lang` folder
3) in `config/app.php`, change the `locale ` to the name of your json file


Ps: Nova, automatically handle the translation, so, your code adding the translation is breaking the app. Just do the 3 steps above AND remove your code in `NovaServiceProvider` that the translation will work perfectly.

**Reply by martio** [reply]
@henryavila What about if the `locale` is English but the admin panel is in only Polish?

**Reply by martio** [reply]
config/app.php

```
'locale' => 'pl',
```

app/Providers/NovaServiceProvider.php 

```
Nova::serving(function (ServingNova $event) {
     app()->setLocale('pl');
});
```

is bad...

<img width="193" alt="Zrzut ekranu 2021-03-8 o 12 23 37" src="https://user-images.githubusercontent.com/133636/110315512-a8ccff00-8009-11eb-94c2-6c46b7ad02cb.png">

**When I replace it with "en" it's fine...**

config/app.php

```
'locale' => 'en',
```

<img width="193" alt="Zrzut ekranu 2021-03-8 o 12 29 29" src="https://user-images.githubusercontent.com/133636/110315826-1b3ddf00-800a-11eb-8c31-be5cbc7be675.png">

Something is wrong...

I'm using translation from: resources/lang/vendor/nova/en.json and resources/lang/vendor/nova/pl.json...

**Reply by crynobone** [reply]
Using `Nova::serving()` is too late. You should have just triggered it without waiting for Nova to loads or using.

```php
Nova::booted(function () {
    app()->setLocale('pl');
});
```

**Reply by martio** [reply]
@crynobone Same. Not everything is translated... Translation files are fine.

**Reply by crynobone** [reply]
What's not translated? Can you be more specific?

**Reply by martio** [reply]
@crynobone eg. "Resources" is not translated in the sidebar... and "Create :resource" or "Update :rescource".

<img width="607" alt="Zrzut ekranu 2021-03-8 o 13 48 04" src="https://user-images.githubusercontent.com/133636/110323470-f00cbd00-8014-11eb-9eb2-ffa8b6193356.png">

<img width="193" alt="Zrzut ekranu 2021-03-8 o 13 47 40" src="https://user-images.githubusercontent.com/133636/110323486-f438da80-8014-11eb-84e9-44b1ad32e749.png">

**Reply by martio** [reply]
The translation file is fine...

**Reply by martio** [reply]
It's weird. When the two locales differ, everything is fine...

Eg.:

config/app.php

```
'locale' => 'de',
```

and

app/Providers/NovaServiceProvider.php

```
Nova::booted(function () {
    app()->setLocale('pl');
});
```

**Reply by crynobone** [reply]
![image](https://user-images.githubusercontent.com/172966/110326597-d5a60400-8053-11eb-8224-130ee4f4db77.png)

Definitely cannot replicate your issue. This definitely not a bug with Nova but I definitely thing something wrong with how you're using it.

**Reply by martio** [reply]
@crynobone Try to set it up like that:

config/app.php

```
'locale' => 'pl',
```

and

app/Providers/NovaServiceProvider.php

```
Nova::booted(function () {
    app()->setLocale('pl');
});
```

Translation in: `resources/lang/vendor/nova/pl.json`

**Reply by crynobone** [reply]
> It's weird. When the two locales differ, everything is fine...

This is irrelevant. Laravel Nova only loads the current locale file. So either `app.locale` set as `en`, or `foo`. It only load the current locale based from `app()->getLocale()`.

**Reply by martio** [reply]
If I give two the same locale, it does not work. How different it works..

**Reply by crynobone** [reply]
![image](https://user-images.githubusercontent.com/172966/110327292-bd82b480-8054-11eb-9825-9eb1934480bb.png)

> Try to set it up like that:

Still getting the same result, there no bug.

> If I give two the same locale, it does not work. How different it works..

This is a usage issue. Something you're doing in your app causing it not to work properly. You need to review what your doing, what packages you include that can cause this issue etc.

**Reply by martio** [reply]
@crynobone Look here...

https://youtu.be/CBCM7l0GxJw

**Reply by crynobone** [reply]
@martio I have tried multiple times to reproduce the issue on a default Nova installation and verify that there no bug in the current implementation. I also have tried to explain to you that you need to check your usage that could cause an issue.

You can show me video, screenshot etc but at the end of the day, it's not reproducable. If you can prove that an issue can  reproduce please submit a failing code to https://github.com/nova-issues/issue-3239/

**Reply by martio** [reply]
@crynobone I found a problem! The problem occurs when usign translations in the JetstreamServiceProvider class, in configurePermissions class. It's not the Nova's fault. Sorry! :(

**Reply by msucevan** [reply]
@crynobone in 
`php 
Nova::booted(function () {
    auth()->user()->language->code;
});
`
return null because is too early to call the authenticated user

**Comment by martio** [comment]
If the localization is the same as in the provider, there are translation problems...

config/app.php => 'en' and NovaServiceProvider => 'de'

<img width="193" alt="Zrzut ekranu 2021-03-8 o 12 56 21" src="https://user-images.githubusercontent.com/133636/110318543-e6338b80-800d-11eb-9eb2-50965c2f881d.png">

config/app.php => 'de' and NovaServiceProvider => 'de'

<img width="193" alt="Zrzut ekranu 2021-03-8 o 12 56 02" src="https://user-images.githubusercontent.com/133636/110318764-30b50800-800e-11eb-8ebf-6f335ba7fae0.png">


NovaServiceProvider.php

```
        Nova::serving(function (ServingNova $event) {
            app()->setLocale('de');
        });
```

**Reply by henryavila** [reply]
@martio, please, do the 3 steps I described **and** remove your code from `app/Providers/NovaServiceProvider.php`. It should work.

make sure you have the file `resources/lang/pl.json`

**Comment by martio** [comment]
I checked. I have only this:

```
    'locale' => 'pl',
```

and "Resources" in the sidebar is not translated...

**Comment by f-liva** [comment]
I've note JetStream and I still suffer this issue. I don't think JetStream
it's the point of fault, there should be an issue with Nova. Everything
works fine with translations on Laravel, but not within Nova backed. Can't
understand why

Il lun 8 mar 2021, 18:10 Marcin Lewandowski <notifications@github.com> ha
scritto:

**Comment by martio** [comment]
@crynobone @henryavila

I think the problem is with Laravel Nova. See:

1. Put Polish translation to:

`resources/lang/vendor/nova/pl.php `

2. Set locale in config/app.php

`'locale' => 'pl',`

3. Set in app/Providers/NovaServiceProvider.php

```
Nova::booted(function () {
    app()->setLocale('pl');
});
```

_Up to this point is fine..._

4. Use translation anywhere in providers...

E.g:

```
class NovaServiceProvider extends NovaApplicationServiceProvider
{
    public function boot()
    {
        parent::boot();

        Nova::booted(function () {
            app()->setLocale('pl');
        });

        __('messages.welcome');
    }
}
```

and... translation in admin panel is incomplete...

**If I remove what is below it is fine:**

**__('messages.welcome');**

**Reply by crynobone** [reply]
```php
    public function boot()
    {
        parent::boot();

        Nova::booted(function () {
            app()->setLocale('pl');
        });

        __('messages.welcome');
    }
```

If you understand the process, you would know how this going to cause issue. `__('messages.welcome')` will be executed way before `Nova::booted()` (because `booted()` will only be executing when nova URI  get executed. 

Not a bug with Laravel Nova, that just the way you use it causing issue.

**Reply by martio** [reply]
@crynobone This is an example from Jetstream package. If you are using Laravel Jetstream, the Laravel Nova does not work properly...

app/Providers/JetstreamServiceProvider.php

```
        Jetstream::role('editor', __('Editor'), [
            'read',
            'create',
            'update',
        ])->description(__('Editor users have the ability to read, create, and update.'));
```

If I remove the translations from the JetstreamServiceProvider class, it works fine.

**How to solve this problem?**

**Reply by crynobone** [reply]
Why are you waiting for Nova booted event to switch locale? That's the whole issue due to your usage. It can't be fixed by Nova or Jetstream

**Reply by martio** [reply]
@crynobone I set the language by user settings:

```
    public function boot()
    {
        parent::boot();

        Nova::serving(function (ServingNova $event) {
            app()->setLocale(
                (('pl' == $event->request->user()->locale) ? 'pl' : 'en')
            );
        });
    }
```

The application has 8 languages and the admin panel only 2.

**Comment by ekandreas** [comment]
I can confirm the same thing in Swedish.

**Comment by msucevan** [comment]
@ekandreas, @martio, @crynobone 

I think here the best solution is to create a middleware and put it in nova.php before HandleInertiaRequests.
```nova.php
    'middleware' => [
        'web',
        SwitchLanguage::class,
        HandleInertiaRequests::class,
  ],
```

This because if you use Nova::serving some translation are translated others not, so I think is a little too late.

**Reply by crynobone** [reply]
Middleware are executed after service providers have been booted. It doesn't solve the issue described by @martio.

**Reply by msucevan** [reply]
Yes it solves it. Because I had the same problem.


When I used in the boot function Nova::serving translation was translating only a few labels. When I switch to use the middleware in nova.php I didn't have any problem.
